File: apple/common/Crypto.swift
import Foundation

class Crypto {

    private let cellSeal: TSCellSeal
    private let localPublicKey: Data
    private let localPrivateKey: Data
    private var peers = [String:Peer]()

    init(password: String) {
        let key = password.data(using: .utf8)!
        cellSeal = TSCellSeal(key: key)

        let keyGeneratorEC: TSKeyGen = TSKeyGen(algorithm: .EC)!
        localPrivateKey = keyGeneratorEC.privateKey as Data
        localPublicKey = keyGeneratorEC.publicKey as Data
    }

    func keyDerivationEncrypt(data: Data) -> Data? {
        do {
            return try cellSeal.wrap(data, context: nil)
        } catch let error as NSError {
            print(error.localizedDescription)
            return nil
        }
    }

    func keyDerivationDecrypt(ciphertext: Data) -> Data? {
        do {
            return try cellSeal.unwrapData(ciphertext, context: nil)
        } catch let error as NSError {
            print("Error occurred while decrypting \(error)", #function)
            return nil
        }
    }

//    func establishSession(forPeerId peerId: String) {
//        let peer = self.peer(forPeerId: peerId)
//        if peer.clientPublicKey == nil {
//            peer.sendPublicKey(isResponse: false)
//        }
//    }
//
    func isSessionEstablishedFor(_ peerId: String) -> Bool {
        let peer = self.peer(forPeerId: peerId)
        if peer.status == .begun {
            peer.sendPublicKey(isResponse: false)
        }
        return peer.status == .sessionEstablished
    }

    private func peer(forPeerId peerId: String) -> Peer {
        if peers[peerId] == nil {
            let peer = Peer(peerId: peerId)
            peers[peerId] = peer
        }
        return peers[peerId]!
    }

    func setPublicKey(key: Data, peerId: String, isResponse: Bool) {
        let peer = self.peer(forPeerId: peerId)
        peer.setServerPublicKey(key: key, isResponse: isResponse)
    }

    func didReceivePayload(_ payload: Data, from peerId: String) {
        peer(forPeerId: peerId).didReceive(payload)
    }

    func encrypt(data: Data, forPeerId peerId: String) -> Data? {
        let peer = self.peer(forPeerId: peerId)
        return peer.encrypt(data)
    }
}

private class Transport: TSSessionTransportInterface {
    private var serverId: String?
    private var serverPublicKeyData: Data?

    func setupKeys(_ serverId: String, serverPublicKey: Data) {
        self.serverId = serverId
        serverPublicKeyData = serverPublicKey
    }

    override func publicKey(for binaryId: Data!) throws -> Data {
        if serverId!.data(using: .utf16)! != binaryId {
            print("mismatch")
        } else {
            print("retrieved public key for \(String(describing: String(data: binaryId!, encoding: String.Encoding.utf16)!))")
        }
        return serverPublicKeyData!
    }
}

private class Peer {

    enum Status {
        case begun
        case publicKeySent
        case sessionEstablished
    }
    var status: Status = .begun

    var transport = Transport()
    var session: TSSession?
    var clientIdData: Data
    var peerId: String
    var clientPrivateKey: Data? = nil
    var clientPublicKey: Data? = nil

    init(peerId: String) {
        clientIdData = Backend.shared.credential!.username.data(using: .utf8)!
        self.peerId = peerId

        guard let keyGeneratorEC: TSKeyGen = TSKeyGen(algorithm: .EC) else {
            print("Error occurred while initialising object keyGeneratorEC", #function)
            return
        }
        clientPrivateKey = keyGeneratorEC.privateKey as Data
        clientPublicKey = keyGeneratorEC.publicKey as Data
    }

    func setServerPublicKey(key: Data, isResponse: Bool) {
        transport.setupKeys(peerId, serverPublicKey: key)
        session = TSSession(userId: clientIdData, privateKey: clientPrivateKey, callbacks: transport)
        if isResponse {
            connect()
        } else {
            sendPublicKey(isResponse: true)
        }
    }

    func sendPublicKey(isResponse: Bool) {
        status = .publicKeySent
        Backend.shared.sendPublicKey(clientPublicKey!, to: peerId, isResponse: isResponse)
    }

    func connect() {
        do {
            guard let message = try session?.connectRequest() else {
                print("could not connectRequest")
                return
            }
            Backend.shared.sendHandshake(message: message, to: peerId)
        } catch {
            print(error.localizedDescription)
        }
    }

//    func sendMessage(_ message: Data) {
//        Backend.shared.sendPayload(message, to: peerId)
//    }

    func didReceive(_ data: Data) {
        print("status is \(status)")
        do {
            let decryptedMessage = try session!.unwrapData(data)
            if !session!.isSessionEstablished() { // themis says: send this back
                print("themis says: send this back")
                Backend.shared.sendHandshake(message: decryptedMessage, to: peerId)
            } else if status != .sessionEstablished { // themis says: session now established
                print("themis says: session now established")
                didEstablishSession(sendThisToo: decryptedMessage)
            } else { // themis says: here is the decrypted message
                print("themis says: here is the decrypted message")
                Backend.shared.didReceiveData(decryptedMessage)
            }
        } catch {
            if let session = session, session.isSessionEstablished() {
                print("themis says: session now established 2")
                didEstablishSession() // themis says: session now established (it can happen this way too)
            } else {
                print(error.localizedDescription)
            }
        }
    }

    private func didEstablishSession(sendThisToo: Data? = nil) {
        status = .sessionEstablished
        if let message = sendThisToo {
            Backend.shared.sendHandshake(message: message, to: peerId)
        }
        Backend.shared.handshook(with: peerId)
    }

    func encrypt(_ data: Data) -> Data? {
        do {
            guard let wrappedMessage: Data = try self.session?.wrap(data) else {
                print("Error occurred during wrapping message ", #function)
                return nil
            }
            return wrappedMessage
        } catch let error as NSError {
            print("Error occurred while wrapping message \(error)", #function)
            return nil
        }
    }
}

File: android/app/src/main/java/red/tel/chat/Crypto.java
package red.tel.chat;

import android.util.Log;

import com.cossacklabs.themis.ISessionCallbacks;
import com.cossacklabs.themis.KeyGenerationException;
import com.cossacklabs.themis.Keypair;
import com.cossacklabs.themis.KeypairGenerator;
import com.cossacklabs.themis.PrivateKey;
import com.cossacklabs.themis.PublicKey;
import com.cossacklabs.themis.SecureCell;
import com.cossacklabs.themis.SecureCellData;
import com.cossacklabs.themis.SecureSession;

import java.io.UnsupportedEncodingException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Crypto {

    private static final String TAG = "Crypto";
    private SecureCell cell;
    private String clientId;
    private PrivateKey clientPrivateKey;
    private PublicKey clientPublicKey;
    private Map<String, Peer> peers = new HashMap<>();

    Crypto(String username, String password) throws UnsupportedEncodingException, KeyGenerationException {
        clientId = username;
        cell = new SecureCell(password);
        Keypair pair = KeypairGenerator.generateKeypair();
        clientPrivateKey = pair.getPrivateKey();
        clientPublicKey = pair.getPublicKey();
    }

    private Peer getPeer(String peerId) {
        if (!peers.containsKey(peerId)) {
            Peer peer = new Peer(peerId);
            peers.put(peerId, peer);
        }
        return peers.get(peerId);
    }

    Boolean isSessionEstablishedFor(String peerId) {
        Peer peer = getPeer(peerId);
        if (peer.status == Status.BEGUN) {
            peer.sendPublicKey(false);
        }
        return peer.status == Status.SESSION_ESTABLISHED;
    }

    byte[] keyDerivationEncrypt(byte[] data) throws Exception {
        SecureCellData cellData = cell.protect(TAG, data);
        return cellData.getProtectedData();
    }

    byte[] keyDerivationDecrypt(byte[] data) throws Exception {
        SecureCellData cellData = new SecureCellData(data, null);
        return cell.unprotect(TAG, cellData);
    }

    void setPublicKey(byte[] key, String peerId, Boolean isResponse) throws Exception {
        Peer peer = getPeer(peerId);
        peer.setServerPublicKey(key, isResponse);
    }

    void onReceivePayload(byte[] payload, String peerId) throws Exception {
        getPeer(peerId).didReceive(payload);
    }

    byte[] encrypt(byte[] data, String peerId) throws Exception {
        return getPeer(peerId).encrypt(data);
    }

    /////// Transport

    private class Transport implements ISessionCallbacks {

        private String peerId;
        private PublicKey serverPublicKey;

        void setupKeys(String peerId, byte[] serverPublicKey) {
            this.peerId = peerId;
            this.serverPublicKey = new PublicKey(serverPublicKey);
        }

        @Override
        public PublicKey getPublicKeyForId(SecureSession session, byte[] id) {
            if (Arrays.equals(id, peerId.getBytes())) {
                return serverPublicKey;
            } else {
                Log.e(TAG, "key id mismatch");
                return null;
            }
        }

        @Override
        public void stateChanged(SecureSession session) {
            // todo: update UI: for example, draw a nice padlock indicating to the user that his/her communication is now secured
            Log.d(TAG, "Peer " + peerId + " state changed to " + session);
        }
    }

    /////// Peer

    private enum Status {
        BEGUN,
        PUBLIC_KEY_SENT,
        SESSION_ESTABLISHED
    }

    private class Peer {

        Status status  = Status.BEGUN;

        private Transport transport = new Transport();
        private SecureSession session;
        private String peerId;

        Peer(String peerId) {
            this.peerId = peerId;
        }

        void setServerPublicKey(byte[] key, Boolean isResponse) throws Exception {
            transport.setupKeys(peerId, key);
            session = new SecureSession(clientId, clientPrivateKey, transport);
            if (isResponse) {
                connect();
            } else {
                sendPublicKey(true);
            }
        }

        private void sendPublicKey(Boolean isResponse) {
            status = Status.PUBLIC_KEY_SENT;
            Backend.shared().sendPublicKey(clientPublicKey.toByteArray(), peerId, isResponse);
        }

        private void connect() {
            try {
                byte[] connectRequest = session.generateConnectRequest();
                Backend.shared().sendHandshake(connectRequest, peerId);
            } catch (Exception exception) {
                Log.e(TAG, exception.getLocalizedMessage());
            }
        }

        void didReceive(byte[] receiveBuffer) throws Exception {
            SecureSession.UnwrapResult result = session.unwrap(receiveBuffer);
            if (session.isEstablished()) {
                status = Status.SESSION_ESTABLISHED;
                Backend.shared().handshook(peerId);        }
            switch (result.getDataType()) {
                case USER_DATA:
                    // this is the actual data that was encrypted by your peer using SecureSession.wrap
                    // process the data according to your application's flow for incoming data
                    Backend.shared().onReceiveData(result.getData());
                    break;
                case PROTOCOL_DATA:
                    // this is the internal Secure Session protocol data. An opaque response was generated, just send it to your peer
                    // send the data to your peer as is
                    Backend.shared().sendHandshake(result.getData(), peerId);
                    break;
                case NO_DATA:
                    // this is the internal Secure Session protocol data, but no response is needed (this usually takes place on the client side when protocol negotiation completes)
                    // do nothing
                    break;
            }
        }

        byte[] encrypt(byte[] data) throws Exception {
            return session.wrap(data);
        }
    }
}
